### DISCUSSION TITLE
I accidentally force pushed to my repo

### DISCUSSION THREAD

**Discussion by Ben3eeE** [community/community#23242] [New to GitHub]
<p>I accidentally force pushed to my repo, and I want to go back to the previous version. What do I do?</p>

**Comment by a-a-ron** [comment]
<p>The ‘git revert’ option undoes a commit by creating another commit. This is important if your repo is being worked on with other collaboraters becasue it doesn’t alter the commit history by re-writing the commit you want to remove.</p>
<p>If you just pushed your changes and you want to go back to the previous version, the following revert command will make that change for you.</p>
<pre><code>git revert HEAD~1
git push origin master
</code></pre>
<p>The above is saying that you want to revert the changes to HEAD by 1, meaning the last commit, make a new commit that undoes those changes, and then push this new commit to the origin branch, in this case the master branch.</p>

**Reply by Subhamstar** [reply]
fatal: bad revision 'HEAD~1'
error: src refspec master does not match any
error: failed to push some refs to 'https://github.com/Subhamstar/JS.git'

**Reply by Subhamstar** [reply]
very helpful

**Comment by lee-dohm** [comment, marked as answer]
<p>If I’m understanding correctly, you’re talking about a scenario like the following:</p>
<ol>
<li>You have a repo named ben3eee/some-repo</li>
<li>You create a few commits on branch some-branch</li>
<li>You push some-branch to ben3eee/some-repo on GitHub using git push</li>
<li>You <a href="https://help.github.com/articles/using-git-rebase-on-the-command-line/" rel="nofollow noopener">squash the commits</a> into one using git rebase -i</li>
<li>You  <strong>force push</strong> some-branch to ben3eee/some-repo on GitHub using git push -f</li>
<li>You now want to restore some-branch to the way it was before step <span class="hashtag">#4</span>
</li>
</ol>
<p>The great thing about Git though is that it does it’s very best to never lose data, so the version of the repository before step <span class="hashtag">#4</span> is still available as long as too much time hasn’t passed. The exact definition of “too much time” is kind of fuzzy and depends on how many other changes you’ve made between step <span class="hashtag">#4</span> and step <span class="hashtag">#6</span>. For this example, I’m going to assume that you realize your mistake right away and no other actions were taken other than the ones in the list above.</p>
<p>The command that can help is called <a href="https://git-scm.com/docs/git-reflog" rel="nofollow noopener">git reflog</a>. You can enter git reflog and see all of the actions that have changed your local repository, including switching branches and rebases, going back quite a ways. I’ve created a simple reflog that shows the scenario I described above:</p>
<pre><code>b46bfc65e (HEAD -&gt; test-branch) HEAD@{0}: rebase -i (finish): returning to refs/heads/test-branch
b46bfc65e (HEAD -&gt; test-branch) HEAD@{1}: rebase -i (squash): a
dd7906a87 HEAD@{2}: rebase -i (squash): # This is a combination of 2 commits.
a3030290a HEAD@{3}: rebase -i (start): checkout refs/heads/master
0c2d866ab HEAD@{4}: commit: c
6cab968c7 HEAD@{5}: commit: b
a3030290a HEAD@{6}: commit: a
c9c495792 (origin/master, origin/HEAD, master) HEAD@{7}: checkout: moving from master to test-branch
c9c495792 (origin/master, origin/HEAD, master) HEAD@{8}: pull: Fast-forward
</code></pre>
<p>You can see at HEAD@{7} I performed a checkout moving from master to test-branch. I then created three commits, “a”, “b” and “c”. Then I rebased them, arriving at HEAD@{0}. The notation HEAD@{number} is the position of HEAD at “number” changes ago. So HEAD@{0} is HEAD where HEAD is now and HEAD@{4} is HEAD four steps ago. We can see from the reflog above that HEAD@{4} is where we need to go in order to restore the branch to where it was before the rebase and 0c2d866ab is the commit ID (also sometimes called “SHA”) for that commit.  So in order to restore test-branch to the state we want, we can issue the command:</p>
<pre><code>git reset --hard HEAD@{4}
</code></pre>
<p>Then we can force push again to restore the repository on GitHub to where it was before.</p>

**Reply by Subhamstar** [reply]
its work

**Reply by Daniel-Georgiev** [reply]
helped me alot when I ws in a very tight situation

**Comment by Ben3eeE** [comment]
<p>This worked! Thank you so much for you help!</p>

**Reply by Subhamstar** [reply]
its work thank you

**Comment by ghost** [comment]
<p>Wouldn’t “git revert” work better in this case if it’s only him altering his repo and not multiple people?</p>
<pre><code>git revert
</code></pre>

**Comment by Zerophase** [comment]
<p>This is not effective at all for rebasing.</p>

**Comment by Zerophase** [comment]
<p>Would that work if there was a rewriting of history involved? revert can work for the simple case, but force commits usually rewrite history and do much more complicated changes.</p>

**Comment by tareq97-zz** [comment]
<p>I accidently pushed a brach in another owner repo. Now i am getting the same message. Question is whether whole team will be notified by my changes or not. Or this message is displayed only my username.</p>

**Comment by hkmaly** [comment]
I've found the hashes of commits involved on https://github.com/[account]/[repository]/activity ... then I did
* git fetch origin [hash]
* git checkout [hash]
* git branch -f master
* git checkout master
* git push origin -f

**Reply by ngothanhthien** [reply]
thanks Hkmaly! It work for me

**Reply by deepakshroff** [reply]
git fetch origin [hash]
git checkout [hash]
git branch -f master
git checkout master
git push origin -f

**Reply by Subhamstar** [reply]
its work thankyou

**Reply by pillo79** [reply]
+1, should be the accepted answer, this works even when the commit was never in the local repo!
